eclipse/eclipse.jdt.ls
# Support Android project

Signed-off-by: Shi Chen <chenshi@microsoft.com>

fix #923, https://github.com/redhat-developer/vscode-java/issues/396

related to #606, #871, https://github.com/redhat-developer/vscode-java/issues/1876, https://github.com/redhat-developer/vscode-java/issues/626, https://github.com/redhat-developer/vscode-java/issues/1390, https://github.com/redhat-developer/vscode-java/issues/1547

requires https://github.com/redhat-developer/vscode-java/pull/2644

This PR adds an init script to help buildship resolve android source path/classpath and add them to eclipse model in Gradle.

The init script contains a Gradle plugin, which will do the following things:

- For Gradle version `5.4` and above, set synchronizationTasks of eclipseModel to automatically generate R/buildConfig/dataBinding files, the following steps will add them to classpath, which makes the android project imported without any manually steps
- For Gradle versions under `5.4`, since the API `synchronizationTasks` is not available, a manual step is required: run Gradle compile Task to generate R/buildConfig/dataBinding files, e.g., `compileDebugSources`, but the plugin will create the empty destination folder and add them to classpath, once the files are created, they will be recognized by the language server.
- Remove standard JDK container since android project should use embedded JDK types included in android.jar
- Add supported source sets to source folders of eclipse model (See `AddSourceFoldersAction`)
  - will add `main`, `test`, `androidTest` by default
- Add dataBinding files to source folders of eclipse model (See `AddDataBindingFilesAction`)
- Add R files to source folders of eclipse model (See `AddRFilesAction`)
- Add buildConfig files to source folders of eclipse model (See `AddBuildConfigFilesAction`)
- Add android.jar to project classpath of eclipse model (See `AddAndroidSDKAction`)
- Add project dependencies to project classpath of eclipse model (See `AddProjectDependenciesAction`)
  - for aar dependencies, extract classes.jar from them to `buildDir/aarLibraries`

> Note: To import Android projects, the specific Android SDK **or** Android SDK manager (recommended) is required, and should be accessible via `ANDROID_HOME` in PATH.
